### PR TITLE
Add new relocation records for block frequency

### PR DESCRIPTION
**Pull request by ashu-mehra**
Created a new relocation records for block frequency and the flag
for checking if the method is already in recompilation queue.

Signed-off-by: Ashutosh Mehra <asmehra@redhat.com>

### CODE REVIEW

**Comment by ashu-mehra** [comment]
fyi - @mpirvu @andrewcraik @r30shah @dsouzai

**Comment by ashu-mehra** [comment]
I have a draft PR in OMR to make use of the new relocation record: https://github.com/eclipse/omr/pull/5202

**Comment by dsouzai** [comment]
Probably worth making this a WIP since this can't be merged until https://github.com/eclipse/omr/pull/5202 is (because of the `TR_BlockFrequency` enum)

**Comment by ashu-mehra** [comment]
> Probably worth making this a WIP since this can't be merged until eclipse/omr#5202 is (because of the TR_BlockFrequency enum)

Right, this is dependent on eclipse/omr#5202.

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:581]: Missing a `break;` here.

Also you should add code to dump the relo data further below in `J9::AheadOfTimeCompile::dumpRelocationHeaderData`

File: runtime/compiler/runtime/RelocationRecord.cpp
Comment (dsouzai) [inline runtime/compiler/runtime/RelocationRecord.cpp:809]: You should probably put a NULL check for the `bodyInfo`. If in some strange circumstance, the compiler in the load run might decide that a particular method shouldn't be recompiled, there won't be a bodyInfo. It would be a truly weird circumstance (especially since in the compile run there was a bodyInfo), so you could add some logging.

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:571]: Probably best to also just fail the compile at this point, there was a new exception recently added `J9::AOTRelocationRecordGenerationFailure` that you could use.

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:575]: Same comment as above.

**Comment by mpirvu** [comment]
@ashu-mehra Hi Ashu, for planning purposes I was wondering about your availability to continue to drive this work item. If job responsibilities prevent you from making progress we need to find someone else to continue the work you started.
Thanks

**Comment by ashu-mehra** [comment]
@mpirvu I have been working on this item. Current status is that the block frequency profiling seems to be working fine on x86 with jitserver. I am looking at the changes required for creating relocation records for ppc which seems to be slightly different than x86.

**Comment by ashu-mehra** [comment]
I have extracted out the code changes required in OMR for this PR into https://github.com/eclipse/omr/pull/5733.
Once that gets merged, this PR should also be ready to go in.
It no longer depends on https://github.com/eclipse/omr/pull/5202.

**Comment by ashu-mehra** [comment]
eclipse/omr#5733 is merged, so  this PR is ready now.

File: runtime/compiler/runtime/RelocationRecord.cpp
Comment (dsouzai) [inline runtime/compiler/runtime/RelocationRecord.cpp:822]: Is it possible for `profileInfo` / `profileInfo->getBlockFrequenceyInfo` to be NULL? If it is then you'll need a NULL check in `applyRelocation` and return non-zero if `_addressToPatch` is NULL; otherwise you'll run into a SIGSEGV.

File: runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp:181]: Why can't this be in the common `runtime/compiler/codegen/J9AheadOfTimeCompile.cpp` file?

File: runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp:297]: You should use `sizeof(TR_RelocationRecordBinaryTemplate)` like you did in `RelocationRecord.cpp`.

**Review by dsouzai — CHANGES_REQUESTED** [review]
Requested some changes. Also:

* You should add the relo record sizes to the tables in all the `J9AheadOfTimeCompile.cpp` files (though if https://github.com/eclipse/openj9/pull/11235 gets merged before this one then it won't matter anymore)
* You need to increase the JITServer minor version

File: runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp
Comment (ashu-mehra) [reply runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp:181]: Only reason is I need to set relo flags in p and not in x, so they are slightly different. I think I can use the implementation for p  as the common code and when generating the relocation I can set relo flag to 0 in x.

Comment (dsouzai) [reply runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp:181]: Yeah it makes sense that p had its own implementation, but the x implementation was the common one, so it could be moved to the common file. The way the infra works is that it will look at the platform specific implementation first, so even if you move the x implementation to the common file, on P it will still run the P implementation.

That said, if you change all platforms to use `TR_RelocationRecordInformation` then you won't need a separate version for P.

Comment (ashu-mehra) [reply runtime/compiler/x/codegen/J9AheadOfTimeCompile.cpp:181]: > if you change all platforms to use TR_RelocationRecordInformation then you won't need a separate version for P

yup, that's the plan.

**Comment by ashu-mehra** [comment]
> You should add the relo record sizes to the tables in all the J9AheadOfTimeCompile.cpp files

`arm/codegen/J9AheadOfTimeCompile.cpp` seems to be missing lot many entries in the table, so I have not updated that.

> You need to increase the JITServer minor version

I guess this will be needed when I enable the changes for jitserver. Currently the jitserver is not doing profiling compilation, so this code should never get executed.

**Comment by ashu-mehra** [comment]
I have addressed rest of the comments in the new commit.

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:1109]: There's a local `comp` object you can use; `TR::comp()` is apparently an expensive operation.

**Review by dsouzai — CHANGES_REQUESTED** [review]
Couple of things to be addressed:

1. Could you squash the commits so that there's just one? The first commit adds code to the x implementation, and the second commit moves it to the common one; this would just muddle the history.
2. You'll have to add `compilationAotBlockFrequencyReloFailure` and `compilationAotRecompQueuedFlagReloFailure` to the switch statement in `TR::CompilationInfo::shouldRetryCompilation`, likely grouped with the other aot load failure codes to ensure that we do JIT compile the method if it fails to load.

**Comment by dsouzai** [comment]
> arm/codegen/J9AheadOfTimeCompile.cpp seems to be missing lot many entries in the table, so I have not updated that.

Yeah I guess it's not really a big deal given that https://github.com/eclipse/openj9/pull/11235 cleans it all up, which hopefully should get merged relatively soon (though this PR will cause a merge conflict that'll delay that PR a bit lol)

**Review by dsouzai — APPROVED** [review]
Thanks for the changes @ashu-mehra

**Comment by dsouzai** [comment]
jenkins test sanity.functional all jdk11

**Comment by dsouzai** [comment]
s390 failures are due to the OMR acceptance being blocked by https://github.com/eclipse/openj9/pull/11480; I'll kick them off later after the acceptance has passed.

**Comment by dsouzai** [comment]
jenkins test sanity.functional zlinux,zlinuxxl jdk11

**Pull request merged**
